### PR TITLE
oo mailroom with tests

### PR DESCRIPTION
**Pull request by mattmaeda**

### CODE REVIEW

**Comment by mattmaeda** [comment]
============================================================================================================ test session starts =============================================================================================================
platform darwin -- Python 3.6.2, pytest-3.2.4, py-1.4.34, pluggy-0.4.0 -- /Library/Frameworks/Python.framework/Versions/3.6/bin/python3.6
cachedir: ../../.cache
rootdir: /Users/matt/Documents/projects/uw/IntroPython-2017/students/mattmaeda/session14/mailroom, inifile:
plugins: cov-2.5.1, celery-4.1.0
collected 24 items                                                                                                                                                                                                                            

test_cli.py::test_get_selection PASSED
test_cli.py::test_send_thank_you PASSED
test_cli.py::test_print_report PASSED
test_cli.py::test_quit_program PASSED
test_cli.py::test_main PASSED
test_cli.py::test_bad_send_thank_you PASSED
test_cli.py::test_main_bad_selection PASSED
test_model.py::test_donor_init_none PASSED
test_model.py::test_donor_init PASSED
test_model.py::test_donor_name PASSED
test_model.py::test_add_donation PASSED
test_model.py::test_total_donation PASSED
test_model.py::test_avg_donation PASSED
test_model.py::test_last_donation PASSED
test_model.py::test_last_donation_none PASSED
test_model.py::test_num_donations PASSED
test_model.py::test_db_init PASSED
test_model.py::test_db_init_initial_list PASSED
test_model.py::test_db_load_from_file PASSED
test_model.py::test_db_save_to_file PASSED
test_model.py::test_donors PASSED
test_model.py::test_add_donor PASSED
test_model.py::test_send_letter PASSED
test_model.py::test_donor_report PASSED

---------- coverage: platform darwin, python 3.6.2-final-0 -----------
Name                                                                                                            Stmts   Miss  Cover
-----------------------------------------------------------------------------------------------------------------------------------
/Users/matt/Documents/projects/uw/IntroPython-2017/students/mattmaeda/session14/mailroom/mailroom/__init__.py       4      0   100%
/Users/matt/Documents/projects/uw/IntroPython-2017/students/mattmaeda/session14/mailroom/mailroom/cli.py           41      1    98%
/Users/matt/Documents/projects/uw/IntroPython-2017/students/mattmaeda/session14/mailroom/mailroom/model.py         67      0   100%
__init__.py                                                                                                         0      0   100%
test_cli.py                                                                                                        51      2    96%
test_model.py                                                                                                     103      0   100%
-----------------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                                             266      3    99%


========================================================================================================= 24 passed in 0.17 seconds ==========================================================================================================

**Comment by mattmaeda** [comment]
Now it works:

========================================================================================================= 25 passed in 0.15 seconds ==========================================================================================================
22:17:13[matt@luna:test](master)$pytest --cov=mailroom -vv
============================================================================================================ test session starts =============================================================================================================
platform darwin -- Python 3.6.2, pytest-3.2.4, py-1.4.34, pluggy-0.4.0 -- /Library/Frameworks/Python.framework/Versions/3.6/bin/python3.6
cachedir: ../../.cache
rootdir: /Users/matt/Documents/projects/uw/IntroPython-2017/students/mattmaeda/session14/mailroom, inifile:
plugins: cov-2.5.1, celery-4.1.0
collected 25 items                                                                                                                                                                                                                            

test_cli.py::test_get_selection PASSED
test_cli.py::test_send_thank_you PASSED
test_cli.py::test_print_report PASSED
test_cli.py::test_quit_program PASSED
test_cli.py::test_main PASSED
test_cli.py::test_bad_send_thank_you PASSED
test_cli.py::test_another_bad_send_thank_you PASSED
test_cli.py::test_main_bad_selection PASSED
test_model.py::test_donor_init_none PASSED
test_model.py::test_donor_init PASSED
test_model.py::test_donor_name PASSED
test_model.py::test_add_donation PASSED
test_model.py::test_total_donation PASSED
test_model.py::test_avg_donation PASSED
test_model.py::test_last_donation PASSED
test_model.py::test_last_donation_none PASSED
test_model.py::test_num_donations PASSED
test_model.py::test_db_init PASSED
test_model.py::test_db_init_initial_list PASSED
test_model.py::test_db_load_from_file PASSED
test_model.py::test_db_save_to_file PASSED
test_model.py::test_donors PASSED
test_model.py::test_add_donor PASSED
test_model.py::test_send_letter PASSED
test_model.py::test_donor_report PASSED

---------- coverage: platform darwin, python 3.6.2-final-0 -----------
Name                                                                                                            Stmts   Miss  Cover
-----------------------------------------------------------------------------------------------------------------------------------
/Users/matt/Documents/projects/uw/IntroPython-2017/students/mattmaeda/session14/mailroom/mailroom/__init__.py       4      0   100%
/Users/matt/Documents/projects/uw/IntroPython-2017/students/mattmaeda/session14/mailroom/mailroom/cli.py           41      0   100%
/Users/matt/Documents/projects/uw/IntroPython-2017/students/mattmaeda/session14/mailroom/mailroom/model.py         67      0   100%
__init__.py                                                                                                         0      0   100%
test_cli.py                                                                                                        55      0   100%
test_model.py                                                                                                     103      0   100%
-----------------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                                                             270      0   100%


========================================================================================================= 25 passed in 0.15 seconds ==========================================================================================================

**Comment by PythonCHB** [comment]
Awesome!


-- 
Christopher Barker, PhD

Python Language Consulting
  - Teaching
  - Scientific Software Development
  - Desktop GUI and Web Development
  - wxPython, numpy, scipy, Cython

File: students/mattmaeda/session14/mailroom/mailroom/test/test_cli.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+def test_get_selection():
+    """ test get selection """
+    user_input = ["1"]
+    with mock.patch("builtins.input", side_effect=user_input):
+        option = cli.get_selection()
+        assert option == "1"
+
+
+@mock.patch('sys.stdout', new_callable=StringIO)
```
Comment (PythonCHB) [inline students/mattmaeda/session14/mailroom/mailroom/test/test_cli.py:59]: very nice!

File: students/mattmaeda/session14/mailroom/mailroom/test/test_cli.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+BAD_THANK_YOU_OUTPUT2 = """Invalid amount '0.001' entered.
+
+Dear Matt,
+
+Thank you for your donation of $100.00.
+
+"""
+
+TEST_REPORT_OUTPUT = """
```
Comment (PythonCHB) [inline students/mattmaeda/session14/mailroom/mailroom/test/test_cli.py:36]: ideally, you wouldn't have the full report to test against -- this way, if you change anything about the report -- number of spaces, names in the test dataset, whatever, you'll need to updated this.

Though you could wait 'till that happens to refactor :-)

Comment (PythonCHB) [reply students/mattmaeda/session14/mailroom/mailroom/test/test_cli.py:36]: Also -- making sure the report is correct really should be in your model tests anyway.

**Review by PythonCHB — APPROVED** [review]
really nice job!

That 100% coverage is satisfying, isn't it?

**Pull request merged**